### ISSUE TITLE
question: Datadog- and opentracing-style tracing header interoperability

### ISSUE THREAD

**Issue by chrisforrette** [DataDog/dd-trace-go#1079]
Hello there, I'm looking for a bit of guidance and I'll caveat: I'm not super knowledgeable about tracing instrumentation so some of my terminology might be a bit fuzzy.

So we run a number of services (~20 or so) and we're spinning up a 3rd party/OSS service that uses Zipkin for tracing and can be configured to emit opentracing-style headers. Unfortunately all of our other services use Datadog-style tracing headers and I'm wondering if there's an easy-ish way to support tracing interoperability between these services with different tracing headers.

I've found the `DD_PROPAGATION_STYLE_INJECT` and `DD_PROPAGATION_STYLE_EXTRACT` environment variables which accept `Datdog` and/or `B3` and set propagators here based on those settings: https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/tracer/textmap.go#L131-L134

And I see that `opentracer` has a similar interface here: https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/opentracer/tracer.go#L69-L91 ...but as far as I can tell you can't set it as a propagator.

And fwiw we're using the `net/http` `WrapRoundTripper` to do most of our tracing: https://github.com/DataDog/dd-trace-go/blob/v1/contrib/net/http/roundtripper.go#L80-L92

My guess is that the answer is: no, you can't do that (easily). But I've seen a variety of compositions with Datadog tracers, opentracing clients, etc., so I'm wondering if there's some incantation I haven't seen that might Just Work™ and I'm hoping to get a bit of insight from some folks who would know more definitively.

Thank you!

**Comment by knusbaum** [comment]
In order to use Opentracing with Datadog, Zipkin should be configured to use the Datadog tracer with opentracing, at which point it *should* be emitting Datadog-style headers. If your Zipkin is using Opentracing with some other tracer, I wouldn't expect distributed tracing to work, even if you translate the headers. 

Before going down the road of translating headers, check that Zipkin's opentracer is configured to use Datadog. (I'm not familiar with Zipkin configuration at all, but hopefully that's possible)

If it *is* configured with Datadog and is still emitting non-datadog headers for some reason, it is possible to implement your own [`Propagator`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#Propagator) which can transfer a `SpanContext` in any way you want. Also have a look at [`NewPropagator`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#NewPropagator) to create a propagator that looks for specific headers. Please note that this may have unintended consequences - If you substitute a custom propagator, the usual headers will no longer get picked up and you will have to use your propagator everywhere.

This can be set for both `ddtrace/tracer` and `ddtrace/opentracer` with the [`WithPropagator`](https://pkg.go.dev/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#WithPropagator) `StartOption`.

If you elaborate a little more about your issue, we may be able to give more/better advice.

**Comment by chrisforrette** [comment]
@knusbaum Thanks for getting back to me! It's great to learn about how to create and configure custom propagators.

The main blocker in this case is that the Zipkin tracing client is baked into a 3rd party application so we have limited controls over the instrumentation details. They have a few tracing header options: opentracing, B3, Jaeger, etc., but not Datadog.

I'm actually at this point working on an upstream contribution to add support for Datadog tracing headers. I see in the source code that they are `uint64`, is there any spec anywhere that you know of detailing how Datadog tracing are formatted/used?

**Comment by knusbaum** [comment]
@chrisforrette
We actually accept B3 headers in the tracer by default (https://github.com/DataDog/dd-trace-go/blob/58598b1cd458abaaabea320cb298b8bd4fbea47b/ddtrace/tracer/textmap.go#L293-L368), so if the client can support B3, you might be able to just configure your application that way.

You can also see here how the tracer encodes the ID integers - basically just uses FormatUint to write them out in decimal format (https://github.com/DataDog/dd-trace-go/blob/58598b1cd458abaaabea320cb298b8bd4fbea47b/ddtrace/tracer/textmap.go#L217-L236).

Again, though, if Zipkin is generating its own Non-Datadog traces that aren't getting sent to Datadog, even if you decode the headers and link to that trace, the Zipkin spans won't appear in our system. I'm not sure if Zipkin can be configured to send traces to our backend or not since I am not familiar with Zipkin. 

Basically, for distributed tracing to work, you need 2 things:
1. spans from *all* services involved in the trace need to be sent to Datadog's backend.
2. Spans from those services need to be linked with distributed tracing (the issue we are discussing here)

Through a cursory scan of the `zipkin-go` repo and other Zipkin documentation, I don't see a way to get Zipkin to send traces to Datadog.

**Comment by knusbaum** [comment]
I don't think it matters for this issue, but I thought I should correct myself. I said that we support B3 headers by default, but that is not true.

They need to be configured using the environment variables `DD_PROPAGATION_STYLE_INJECT` and `DD_PROPAGATION_STYLE_EXTRACT`:
https://github.com/DataDog/dd-trace-go/blob/307fc73cc29125423c0f9e0e9295ce659d8fe69d/ddtrace/tracer/textmap.go#L66-L70

By default, only datadog headers are propagated:
https://github.com/DataDog/dd-trace-go/blob/307fc73cc29125423c0f9e0e9295ce659d8fe69d/ddtrace/tracer/textmap.go#L148-L153

**Comment by chrisforrette** [comment]
@knusbaum Thanks! Yeah, I had seen the docs on B3 headers here: https://docs.datadoghq.com/tracing/setup_overview/setup/go/?tab=containers#b3-headers-extraction-and-injection

The problem for us is that we're running this one 3rd party service that we can configure to emit B3 style tracing headers but it does not support Datadog headers, and all of our other services use Datadog headers. So we could configure those other services to inject and extract B3 headers, but we would have to migrate them to _only_ use B3 headers, otherwise calls to the 3rd party service would drop from the trace. And that would be a pretty hefty migration.

In any case, I appreciate all your help! I'll go ahead and close this out.
